File: scripts/bug-search.py
#!/usr/bin/env python3
"""bug-chronicle 搜索。按关键词查找相关 bug，输出结构化摘要。"""
import argparse
import json
import sys
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
BUG_FILE = ROOT / "docs/architecture/bug-chronicle.md"


def load_bugs():
    """将 bug-chronicle 解析为 bug 列表。"""
    text = BUG_FILE.read_text("utf-8", errors="replace")
    bugs = []
    current = None
    current_section = ""

    for line in text.split("\n"):
        m = re.match(r"^## (\d+\.\s*.+)$", line)
        if m:
            if current:
                bugs.append(current)
            current = {"title": m.group(1), "sections": {}}
            current_section = ""
            continue

        m2 = re.match(r"^### (.+)$", line)
        if m2 and current:
            current_section = m2.group(1)
            current["sections"][current_section] = ""
            continue

        if current and current_section:
            current["sections"][current_section] += line + "\n"

    if current:
        bugs.append(current)
    return bugs


def search(keyword, bugs):
    """在 bug 列表中搜索关键词，返回匹配的 bug 子集。"""
    kw = keyword.lower()
    results = []
    for b in bugs:
        full_text = json.dumps(b, ensure_ascii=False).lower()
        if kw in full_text:
            results.append(b)
    return results


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("keyword", nargs="?", help="搜索关键词")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    bugs = load_bugs()

    if args.keyword:
        results = search(args.keyword, bugs)
    else:
        results = bugs

    summary = {"total_bugs": len(bugs), "matched": len(results)}

    if args.json:
        out = {"_summary": summary, "bugs": results}
        sys.stdout.buffer.write(json.dumps(out, ensure_ascii=False, indent=2).encode("utf-8"))
        sys.stdout.buffer.write(b"\n")
    else:
        if not args.keyword:
            print(f"共 {len(bugs)} 条 bug 记录")
            return
        print(f"关键词 '{args.keyword}' 匹配 {len(results)} 条:\n")
        for b in results:
            print(f"  ## {b['title']}")
            for sec, text in b["sections"].items():
                first_line = text.strip().split("\n")[0][:80] if text.strip() else ""
                print(f"    {sec}: {first_line}")


if __name__ == "__main__":
    main()

File: scripts/bug-search.ts
#!/usr/bin/env node
/**
 * bug-chronicle 搜索。按关键词查找相关 bug，输出结构化摘要。
 * 由 scripts/bug-search.py 迁移（2026-08-03），逻辑逐点保真。
 * bug-search.ts — Bug 搜索工具
 * 设计意图：Bug 搜索工具
 * 依赖：node:fs / node:path / node:url
 * 用法：
 *   node scripts/bug-search.ts                 # 默认行为
 *   node scripts/bug-search.ts --json # JSON 输出（CI/子代理消费）
 * 退出码：0（无 process.exit 调用）
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT } from './_lib/scan-files.ts';
import { parseArgs } from './_lib/parse-args.ts';

const BUG_FILE = path.join(ROOT, 'docs/archive/bug-chronicle.md');

function loadBugs() {
  /** 将 bug-chronicle 解析为 bug 列表。 */
  const text = fs.readFileSync(BUG_FILE, 'utf-8');
  const bugs: any[] = [];
  let current: any = null;
  let currentSection = '';

  for (const line of text.split('\n')) {
    const m = line.match(/^## (\d+\.\s*.+)$/);
    if (m) {
      if (current) bugs.push(current);
      current = { title: m[1], sections: {} };
      currentSection = '';
      continue;
    }

    const m2 = line.match(/^### (.+)$/);
    if (m2 && current) {
      currentSection = m2[1]!;
      current.sections[currentSection] = '';
      continue;
    }

    if (current && currentSection) {
      current.sections[currentSection] += line + '\n';
    }
  }

  if (current) bugs.push(current);
  return bugs;
}

function search(keyword: string, bugs: any[]) {
  /** 在 bug 列表中搜索关键词，返回匹配的 bug 子集。 */
  const kw = keyword.toLowerCase();
  const results: any[] = [];
  for (const b of bugs) {
    const fullText = JSON.stringify(b).toLowerCase();
    if (fullText.includes(kw)) results.push(b);
  }
  return results;
}

const args = parseArgs(process.argv.slice(2), { bools: ['json'] });
if (args.unknown.length) {
  console.error(`❌ 未知参数: ${args.unknown.join(', ')}（--help 查看用法）`);
  process.exit(2);
}
const jsonMode = args.json;
const keywordArg = args._[0] ?? null;

const bugs = loadBugs();
const results = keywordArg ? search(keywordArg, bugs) : bugs;
const summary = { total_bugs: bugs.length, matched: results.length };

if (jsonMode) {
  const out = { _summary: summary, bugs: results };
  process.stdout.write(JSON.stringify(out, null, 2) + '\n');
} else {
  if (!keywordArg) {
    console.log(`共 ${bugs.length} 条 bug 记录`);
  } else {
    console.log(`关键词 '${keywordArg}' 匹配 ${results.length} 条:\n`);
    for (const b of results) {
      console.log(`  ## ${b.title}`);
      for (const [sec, text] of Object.entries(b.sections as Record<string, any>)) {
        const trimmed = text.trim();
        const firstLine = trimmed ? trimmed.split('\n')[0].slice(0, 80) : '';
        console.log(`    ${sec}: ${firstLine}`);
      }
    }
  }
}
